### PR TITLE
Clarify when route_service_url is required

### PR DESCRIPTION
**Pull request by duglin**
Clarify when route_service_url is required
    
See: https://github.com/openservicebrokerapi/servicebroker/pull/166/files#r113097378
    
Closes: #169
    
Add some clarity around when a broker can return extra data in the bindings
    
I decided to reword this to be a requirement on the broker. Talking about
the platform didn't make any sense to me since we can't really say anything
normative about it from the POV.
    
While in there I also `s/broker client/platform/g` for consistency.

Signed-off-by: Doug Davis <dug@us.ibm.com>

### CODE REVIEW

File: spec.md [hunk 1]
```diff
@@ -540,11 +540,28 @@ The `requires` field in the [Catalog](#catalog-mgmt) endpoint enables a platform
...
+  routable address, a platform MUST send the address along with the request
+  to create a binding using `"bind_resource":{"route":"some-address.com"}`.
 
-The `requires` field in the [Catalog](#catalog-mgmt) endpoint enables a platform marketplace to validate requests to create bindings. A platform MAY reject requests to create bindings when a broker has declared `"requires":["route_forwarding"]` for a service in the catalog endpoint.
+- Upon receiving an incoming appliation request, the platform needs to know
+  where to send this request to. To facilitate this, the broker MUST return
+  a `route_service_url` in the response for a request to create a binding
+  unless there has been some out-of-band configuration already performed;
+  in which case the broker MUST NOT return a `route_service_url` parameter.
+
```
Comment (angarg12) [inline spec.md]: I would reword this a bit and split it into two bullet points. Something like the following:

- To facilitate platforms to forward application requests, the broker MUST return  a `route_service_url` in the response for a request to create a binding.
- If out-of-band configuration of routes is performed, brokers MUST NOT return a `route_service_url` parameter.

Comment (duglin) [inline spec.md]: I like the idea of splitting the sentence into two.  However, when I moved the 2nd sentence to a new bullet it felt a little odd because those two sentences really need to go hand-in-hand to make sure people realize the MUST has a conditional aspect to it.  see what you think of my edits and let me know if you still think a 2nd bullet instead of a second sentence is still needed.

**Comment by angarg12** [comment]
lgtm

**Comment by duglin** [comment]
LGTM

**Comment by avade** [comment]
LGTM

**Comment by duglin** [comment]
rebase, so LGTM's needed, again

LGTM

**Comment by angarg12** [comment]
LGTM

File: spec.md [hunk 2]
```diff
@@ -534,23 +534,50 @@ Credentials are a set of information used by an application or a user to utilize
 
 #### Log Drain ####
 
-There are a class of service offerings that provide aggregation, indexing, and analysis of log data. To utilize these services an application that generates logs needs information for the location to which it will stream logs. If a broker represents one of these services, it MAY return a `syslog_drain_url` in the response for a request to create a service binding, to which logs MUST be streamed.
+There are a class of service offerings that provide aggregation, indexing, and
+analysis of log data. To utilize these services an application that generates
+logs needs information for the location to which it will stream logs. A
+create binding response from one of these services MUST include a
+`syslog_drain_url`. This URL MUST be used when sending logs to the service.
```
Comment (pmorie) [inline spec.md]: We lost this language:

>  to which logs MUST be streamed.

...which is not equivalent to this sentence:

> This URL MUST be used when sending logs to the service.

Was that intentional?

File: spec.md [hunk 3]
```diff
@@ -534,23 +534,50 @@ Credentials are a set of information used by an application or a user to utilize
...
+There are a class of service offerings that intermediate requests to
+applications, performing functions such as rate limiting or authorization.
+If the platform and service both support this feature then the following
+applies:
 
-Some platforms MAY support proxying of application requests to service instances. In this case the platform needs to know where to send application requests; to facilitate this, the broker MUST return a `route_service_url` in the response for a request to create a binding. Not all services of this type expect to receive requests proxied by the platform; some services will have been configured out-of-band to intermediate requests to applications. In this case, the broker will not return `route_service_url` in response to the create binding request. By sending `bind-resource` as described above, the platform enables dynamic configuration of a service instance already in the application request path for the route, requiring no change in the platform routing tier.
+- To configure a service instance with behavior specific to an application's
```
Comment (pmorie) [inline spec.md]: I think this needs to be related to `route_service_url`

File: spec.md [hunk 4]
```diff
@@ -534,23 +534,50 @@ Credentials are a set of information used by an application or a user to utilize
...
+  routable address, a platform MUST send the address along with the request
+  to create a binding using `"bind_resource":{"route":"some-address.com"}`.
 
-The `requires` field in the [Catalog](#catalog-mgmt) endpoint enables a platform marketplace to validate requests to create bindings. A platform MAY reject requests to create bindings when a broker has declared `"requires":["route_forwarding"]` for a service in the catalog endpoint.
+- Upon receiving an incoming appliation request, the platform needs to know
```
Comment (pmorie) [inline spec.md]: typo: application

File: spec.md [hunk 5]
```diff
@@ -534,23 +534,50 @@ Credentials are a set of information used by an application or a user to utilize
...
+  for the route, requiring no change in the platform routing tier.
+
+Brokers MUST NOT include a `route_service_url` in a create binding response
+if it did not include a `"requires":["route_forwarding"]` property in its
+[Catalog](#catalog-mgmt) endpoint.
 
 #### Volume Services ####
 
-There are a class of services that provide network storage to applications via volume mounts in the application container. A service broker MUST return data needed for this configuration with `volume_mount` in response to the request to create a binding.
+There are a class of services that provide network storage to applications
```
Comment (pmorie) [inline spec.md]: Should tie to `requires` here

File: spec.md (hunk 2 above)
Comment (duglin) [inline spec.md]: In what way do you seem them as being different?

File: spec.md (hunk 3 above)
Comment (duglin) [inline spec.md]: Can you elaborate? What text are you looking for?

File: spec.md (hunk 4 above)
Comment (duglin) [inline spec.md]: fixed

File: spec.md (hunk 5 above)
Comment (duglin) [inline spec.md]: does the sentence below (line 578) cover it?

**Comment by duglin** [comment]
rebased - another round of LGTMs needed

**Comment by duglin** [comment]
rebased

**Comment by avade** [comment]
lgtm

**Comment by angarg12** [comment]
LGTM

**Comment by duglin** [comment]
rebased

**Comment by duglin** [comment]
ready for reviews

File: spec.md (hunk 2 above)
Comment (avade) [inline spec.md]: I think it should say

```
There are a class of service offerings that provide aggregation, indexing, and
analysis of log data. To utilize these services an application that generates 
logs needs information for the location to which it will stream logs. 
A create binding response from a service broker that provides one of these services 
MUST include a `syslog_drain_url`. The platform MUST use `syslog_drain_url` when
sending logs to the service.
```

My only question Is it the platform that streams the logs to this service or the application? @zrob may know..

Comment (zrob) [inline spec.md]: The platform handles sending STDOUT and STDERR to the syslog drain.

Comment (avade) [inline spec.md]: Thanks @zrob.

Ok, I think my suggested language for the works. WDYT @duglin / @pmorie?

Comment (duglin) [inline spec.md]: ok updated - PTAL

File: spec.md [hunk 6]
```diff
@@ -408,7 +408,7 @@ For success responses, a broker MUST return the following fields. For error resp
 | Response field | Type | Description |
 | --- | --- | --- |
 | dashboard_url | string | The URL of a web-based management user interface for the service instance; we refer to this as a service dashboard. The URL MUST contain enough information for the dashboard to identify the resource being accessed (`9189kdfsk0vfnku` in the example below). Note: a broker that wishes to return `dashboard_url` for a service instance MUST return it with the initial response to the provision request, even if the service is provisioned asynchronously. |
-| operation | string | For asynchronous responses, service brokers MAY return an identifier representing the operation. The value of this field SHOULD be provided by the broker client with requests to the [Last Operation](#polling-last-operation) endpoint in a URL encoded query parameter. |
+| operation | string | For asynchronous responses, service brokers MAY return an identifier representing the operation. The value of this field SHOULD be provided by the platform with requests to the [Last Operation](#polling-last-operation) endpoint in a URL encoded query parameter.  |
```
Comment (vaikas-google) [inline spec.md]: L294 above says this MUST be provided by the platform, why is it SHOULD here?

Comment (duglin) [inline spec.md]: hmm, good catch. I'll change this to a MUST.

**Comment by vaikas-google** [comment]
LGTM

**Comment by angarg12** [comment]
LGTM

**Comment by zrob** [comment]
LGTM

**Comment by avade** [comment]
LGTM

**Comment by avade** [comment]
Needs another round @angarg12 @vaikas-google

File: spec.md [hunk 7]
```diff
@@ -519,7 +519,7 @@ For success responses, a broker MUST return the following field. Others will be
 
 ## Binding
 
-If `bindable:true` is declared for a service or plan in the [Catalog](#catalog-management) endpoint, broker clients MAY request generation of a service binding.
+If `bindable:true` is declared for a service or plan in the [Catalog](#catalog-management) endpoint, platform MAY request generation of a service binding.
```
Comment (mattmcneeney) [inline spec.md]: _the_ platform

File: spec.md [hunk 8]
```diff
@@ -531,23 +531,51 @@ Credentials are a set of information used by an application or a user to utilize
...
+There are a class of service offerings that intermediate requests to
+applications, performing functions such as rate limiting or authorization.
+If the platform and service both support this feature then the following
+applies:
 
-Some platforms MAY support proxying of application requests to service instances. In this case the platform needs to know where to send application requests; to facilitate this, the broker MUST return a `route_service_url` in the response for a request to create a binding. Not all services of this type expect to receive requests proxied by the platform; some services will have been configured out-of-band to intermediate requests to applications. In this case, the broker will not return `route_service_url` in response to the create binding request. By sending `bind_resource` as described above, the platform enables dynamic configuration of a service instance already in the application request path for the route, requiring no change in the platform routing tier.
+- To configure a service instance with behavior specific to an application's
```
Comment (mattmcneeney) [inline spec.md]: Is this section supposed to be indented?

Comment (pmorie) [inline spec.md]: It looks like that is the intention to me

File: spec.md [hunk 9]
```diff
@@ -531,23 +531,51 @@ Credentials are a set of information used by an application or a user to utilize
...
+  where to send this request to. To facilitate this, the broker MUST return
+  a `route_service_url` in the response for a request to create a binding.
+  However, if out-of-band configuration was performed then brokers MUST NOT
+  return return a `route_service_url` parameter.
+
+  By sending `bind-resource` as described above, the platform enables dynamic
+  configuration of a service instance already in the application request path
+  for the route, requiring no change in the platform routing tier.
+
+Brokers MUST NOT include a `route_service_url` in a create binding response
```
Comment (pmorie) [inline spec.md]: I think the value of 'it' is ambiguous here, consider changing so it's explicit as the service the instance was provisioned from.

File: spec.md [hunk 10]
```diff
@@ -531,23 +531,51 @@ Credentials are a set of information used by an application or a user to utilize
...
 
 #### Volume Services
 
-There are a class of services that provide network storage to applications via volume mounts in the application container. A service broker MUST return data needed for this configuration with `volume_mount` in response to the request to create a binding.
+There are a class of services that provide network storage to applications
+via volume mounts in the application container. A create binding response
+from one of these services MUST include a `volume_mount`.
 
-The `requires` field in the [Catalog](#catalog-management) endpoint enables a platform marketplace to validate a response for create binding that includes a `volume_mounts`. Platform marketplaces can consider a broker's response invalid if it includes a `volume_mounts` and `"requires":["volume_mount"]` is not present in the [Catalog](#catalog-management) endpoint.
+Brokers MUST NOT include a `volume_mounts` in a create binding response
```
Comment (pmorie) [inline spec.md]: Same issue with 'it'

File: spec.md [hunk 11]
```diff
@@ -732,7 +760,7 @@ For success responses, the following fields are supported. Others will be ignore
 
 | Response field | Type | Description |
 | --- | --- | --- |
-| operation | string | For asynchronous responses, service brokers MAY return an identifier representing the operation. The value of this field MUST be provided by the broker client with requests to the [Last Operation](#polling-last-operation) endpoint in a URL encoded query parameter. If present, MUST be a non-empty string. |
+| operation | string | For asynchronous responses, service brokers MAY return an identifier representing the operation. The value of this field MUST be provided by the platform with requests to the [Last Operation](#polling-last-operation) endpoint in a URL encoded query parameter. If present, MUST be a non-empty string. |
```
Comment (pmorie) [inline spec.md]: 'If preset, the value of this field'

File: spec.md (hunk 7 above)
Comment (duglin) [inline spec.md]: fixed

File: spec.md (hunk 8 above)
Comment (duglin) [inline spec.md]: yea, I couldn't get it to read right a normal text so I tried a "list" and it seemed better.  I'm open to other alternatives.

File: spec.md (hunk 9 above)
Comment (duglin) [inline spec.md]: fixed in all of 'em

File: spec.md (hunk 11 above)
Comment (duglin) [inline spec.md]: "preset" ?

File: spec.md [hunk 12]
```diff
@@ -532,23 +532,51 @@ Credentials are a set of information used by an application or a user to utilize
...
-The `requires` field in the [Catalog](#catalog-management) endpoint enables a platform marketplace to validate requests to create bindings. A platform MAY reject requests to create bindings when a broker has declared `"requires":["route_forwarding"]` for a service in the catalog endpoint.
+- Upon receiving an incoming application request, the platform needs to know
+  where to send this request to. To facilitate this, the broker MUST return
+  a `route_service_url` in the response for a request to create a binding.
+  However, if out-of-band configuration was performed then brokers MUST NOT
+  return return a `route_service_url` parameter.
+
+  By sending `bind-resource` as described above, the platform enables dynamic
+  configuration of a service instance already in the application request path
+  for the route, requiring no change in the platform routing tier.
```
Comment (shalako) [inline spec.md]: This seems redundant, as `bind-resource` is described in the previous bullet.

File: spec.md [hunk 13]
```diff
@@ -532,23 +532,51 @@ Credentials are a set of information used by an application or a user to utilize
...
+- To configure a service instance with behavior specific to an application's
+  routable address, a platform MUST send the address along with the request
+  to create a binding using `"bind_resource":{"route":"some-address.com"}`.
 
-The `requires` field in the [Catalog](#catalog-management) endpoint enables a platform marketplace to validate requests to create bindings. A platform MAY reject requests to create bindings when a broker has declared `"requires":["route_forwarding"]` for a service in the catalog endpoint.
+- Upon receiving an incoming application request, the platform needs to know
+  where to send this request to. To facilitate this, the broker MUST return
+  a `route_service_url` in the response for a request to create a binding.
+  However, if out-of-band configuration was performed then brokers MUST NOT
+  return return a `route_service_url` parameter.
```
Comment (shalako) [inline spec.md]: I feel like some clarity has been lost in the new version. Suggestion:

"Some platforms support proxying of application requests to service instances. If a service is deployed in a configuration to support this behavior, the broker MUST return a `route_service_url` in the response for a request to create a binding, so that the platform knows where to send the application request. If the service is deployed such that the network configuration to proxy application requests through instances of the service is managed out-of-band, the broker MUST NOT return `route_service_url` in the response."

File: spec.md (hunk 8 above)
Comment (shalako) [inline spec.md]: This section reads a bit like sending the `bind-resource` field is optional based on whether the platform supports configuration of the instance specific to the routable address.

As this is up to the broker author, we should simply say that the platform must send `bind-resource` if it supports Route Services. The broker author may choose to support configuration by routable address, or not. A broker could implement service instances that perform the same operation no matter what the address, but that require the route to perform the operation at all.

Suggestion:
"If a platform supports Route Services, it MUST send a routable address or endpoint for the application along with the request to create a service binding using `"bind_resource":{"route":"some-address.com"}`. In the case that a platform supports binding multiple routable addresses to the same service instance, the broker author MAY support configuration specific to an address using `parameters`."

File: spec.md (hunk 12 above)
Comment (shalako) [inline spec.md]: I think we can get rid of this paragraph, as it is a duplicate of a previous bullet.

File: spec.md (hunk 10 above)
Comment (duglin) [inline spec.md]: fixed

**Comment by duglin** [comment]
ping @shalako - see if the route stuff is ok

File: spec.md [hunk 14]
```diff
@@ -531,23 +531,49 @@ Credentials are a set of information used by an application or a user to utilize
...
+If a platform supports route services, it MUST send a routable address,
+or endpoint, for the application along with the request to create a service
+binding using `"bind_resource":{"route":"some-address.com"}`. A broker MAY
+support configuration specific to an address using parameters; exposing this
+feature to users would require a platform to support binding multiple
+routable addresses to the same service instance.
 
-The `requires` field in the [Catalog](#catalog-management) endpoint enables a platform marketplace to validate requests to create bindings. A platform MAY reject requests to create bindings when a broker has declared `"requires":["route_forwarding"]` for a service in the catalog endpoint.
+If a service is deployed in a configuration to support this behavior, the
+broker MUST return a `route_service_url in the response for a request to
```
Comment (shalako) [inline spec.md]: Missing a back tick.

Comment (duglin) [inline spec.md]: fixed

**Comment by shalako** [comment]
LGTM

**Comment by angarg12** [comment]
LGTM

**Comment by vaikas-google** [comment]
LGTM

**Comment by avade** [comment]
LGTM
